src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -279,7 +310,11 @@ private string GenerateId()
             {
                 // Start from outside the process (e.g. incoming HTTP)
                 Debug.Assert(ParentId.Length != 0);
-                ret = appendSuffix(ParentId, "I_" + Interlocked.Increment(ref s_currentRootId));
+
+                //sanitize external RequestId as it may not be hierarchical. 
+                //we cannot update ParentId, we must let it be logged exactly as it was passed.
+                string parentId = ParentId[0] == s_rootIdPrefix ? ParentId : s_rootIdPrefix + ParentId;
```

Comment:
You have an assert for length != 0, but should you check it anyway here to avoid a potential IndexOutOfRangeException?